### PR TITLE
Feature/alarming response

### PR DESCRIPTION
**Pull request by billypchan**
## What's new in this PR?

Draft for Feature/alarming response

### CODE REVIEW

File: Source/Model/ButtonState/ButtonState.swift [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        buttonState.message = message
+        return buttonState
+    }
+    
+    override static func entityName() -> String {
+        return String(describing: ButtonState.self)
+    }
+    
+    override static func isTrackingLocalModifications() -> Bool {
+        return false
```
Comment (David-Henner) [inline Source/Model/ButtonState/ButtonState.swift]: should this be `true` in order to observe changes to the `ButtonState` ? 
(We need to be notified of changes in order to process the button action confirmation in the UI)

File: Source/Model/Message/ZMClientMessage+Encryption.swift [hunk 2]
```diff
@@ -206,6 +215,9 @@ extension ZMGenericMessage {
             }
             recipientUsers = recipients
         }
+        else if self.hasButtonAction() {
```
Comment (billypchan) [inline Source/Model/Message/ZMClientMessage+Encryption.swift]:
```suggestion
        } else if hasButtonAction() {
```

File: Source/Model/Message/ZMClientMessage+Encryption.swift [hunk 3]
```diff
@@ -206,6 +215,9 @@ extension ZMGenericMessage {
...
+            recipientUsers = recipientForButtonActionMessage()
+        }
         else if let deletedEphemeral = recipientsForDeletedEphemeral() {
```
Comment (billypchan) [inline Source/Model/Message/ZMClientMessage+Encryption.swift]:
```suggestion
        } else if let deletedEphemeral = recipientsForDeletedEphemeral() {
```

File: Source/Model/ButtonState/ButtonState.swift (hunk 1 above)
Comment (typfel) [reply Source/Model/ButtonState/ButtonState.swift]: `isTrackingLocalModifications` are not related to the observer system. It controls if changes to properties are tracked in the `modifiedKeys` property, which is used when syncing the object with the backend.

We are not syncing the ButtonState entity to the backend AFAIK so this should be `false`.

File: Source/Model/Message/ZMClientMessage+Encryption.swift [hunk 4]
```diff
@@ -151,6 +151,15 @@ extension ZMGenericMessage {
     func recipientUsersForMessage(in conversation: ZMConversation, selfUser: ZMUser) -> (users: Set<ZMUser>, strategy: MissingClientsStrategy) {
         let (services, otherUsers) = conversation.localParticipants.categorizeServicesAndUser()
 
+        func recipientForButtonActionMessage() -> Set<ZMUser> {
+            guard self.hasButtonAction(),
```
Comment (David-Henner) [inline Source/Model/Message/ZMClientMessage+Encryption.swift]:
```suggestion
            guard hasButtonAction(),
```

File: Source/Model/Message/ZMClientMessage+Composite.swift [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+extension ZMClientMessage: ConversationCompositeMessage {
+    public var compositeMessageData: CompositeMessageData? {
+        guard case .some(.composite) = underlyingMessage?.content else {
+            return nil
+        }
+        return self
+    }
+}
+
+extension ZMClientMessage {
```
Comment (typfel) [inline Source/Model/Message/ZMClientMessage+Composite.swift]: Let's move the button related to code to its own extension?

**Comment by billypchan** [comment]
@zenkins release major

**Pull request merged**